### PR TITLE
Bug fix 3.3/nsis improvements

### PR DESCRIPTION
**Pull request by dothebart**

### CODE REVIEW

**Comment by Simran-B** [comment]
I had 3.3.13 running as service in the background and used a test build of the installer but used the user-mode installation (non-service). This failed, probably because it can't start the server on the same port:

```
---------------------------
Installation von ArangoDB3 3.3.19
---------------------------
Failed to initialize database password

:

Unknown exit code -1073741819!
---------------------------
OK   
---------------------------
```

**Comment by Simran-B** [comment]
Content of `C:\Users\<user>\AppData\Local\ArangoDB3 3.3.19` after failed installation:

```
etc
LICENSE.txt
LICENSES-OTHER-COMPONENTS.md
README.md
README.txt
README.windows.txt
reduceRight.js
reject.js
remove.js
repeat.js
replace.js
resources
rest.js
result.js
reverse.js
round.js
sample.js
sampleSize.js
seq.js
set.js
setWith.js
shuffle.js
size.js
slice.js
snakeCase.js
some.js
sortBy.js
sortedIndex.js
sortedIndexBy.js
sortedIndexOf.js
sortedLastIndex.js
sortedLastIndexBy.js
sortedLastIndexOf.js
sortedUniq.js
sortedUniqBy.js
split.js
spread.js
startCase.js
startsWith.js
string.js
stubArray.js
stubFalse.js
stubObject.js
stubString.js
stubTrue.js
subtract.js
sum.js
sumBy.js
tail.js
take.js
takeRight.js
takeRightWhile.js
takeWhile.js
tap.js
template.js
templateSettings.js
throttle.js
thru.js
times.js
toArray.js
toFinite.js
toInteger.js
toIterator.js
toJSON.js
toLength.js
toLower.js
toNumber.js
toPairs.js
toPairsIn.js
toPath.js
toPlainObject.js
toSafeInteger.js
toString.js
toUpper.js
transform.js
trim.js
trimEnd.js
trimStart.js
truncate.js
unary.js
unescape.js
Uninstall.exe
union.js
unionBy.js
unionWith.js
uniq.js
uniqBy.js
uniqueId.js
uniqWith.js
unset.js
unzip.js
unzipWith.js
update.js
updateWith.js
upperCase.js
upperFirst.js
usr
util.js
value.js
valueOf.js
values.js
valuesIn.js
var
without.js
words.js
wrap.js
wrapperAt.js
wrapperChain.js
wrapperLodash.js
wrapperReverse.js
wrapperValue.js
xor.js
xorBy.js
xorWith.js
zip.js
zipObject.js
zipObjectDeep.js
zipWith.js
```

Why are all these JS files in the root installation folder?

File: Installation/Windows/Templates/NSIS.template.in [hunk 1]
```diff
@@ -205,16 +217,13 @@ Function check_previous_install
 
   IfFileExists "$DATADIR\@INC_CPACK_ARANGO_DATA_DIR@\SERVER" OldFound
   IfFileExists "$DATADIR\@INC_CPACK_ARANGO_DATA_DIR@\ENGINE" OldFound
+  DetailPrint "This is a fresh install."
```
Comment (Simran-B) [inline Installation/Windows/Templates/NSIS.template.in]:
```suggestion
  DetailPrint "This is a fresh installation."
```

File: Installation/Windows/Templates/NSIS.template.in [hunk 2]
```diff
@@ -205,16 +217,13 @@ Function check_previous_install
 
   IfFileExists "$DATADIR\@INC_CPACK_ARANGO_DATA_DIR@\SERVER" OldFound
   IfFileExists "$DATADIR\@INC_CPACK_ARANGO_DATA_DIR@\ENGINE" OldFound
+  DetailPrint "This is a fresh install."
   return
 OldFound:
+  DetailPrint "found old installation."
```
Comment (Simran-B) [inline Installation/Windows/Templates/NSIS.template.in]:
```suggestion
  DetailPrint "Found old installation."
```

File: Installation/Windows/Templates/NSIS.template.in [hunk 3]
```diff
@@ -715,36 +723,44 @@ FunctionEnd
...
+  abortUpgrade:
+    Abort
+  ${EndIf}
+
   ${If} $0 != 0
-  ${AndIf} $0 != 11
-    !insertmacro printExitCode $0 "failed to detect whether we need to Upgrade"
-  ${ElseIf} $0 == 11
+  ${AndIf} $0 != ${ARANGO_EXIT_UPGRADE_REQUIRED}
+    !insertmacro printExitCode $0 "failed to detect whether we need to Upgrade" ""
```
Comment (Simran-B) [inline Installation/Windows/Templates/NSIS.template.in]:
```suggestion
    !insertmacro printExitCode $0 "Failed to detect whether we need to upgrade" ""
```

File: Installation/Windows/Templates/NSIS.template.in [hunk 4]
```diff
@@ -715,36 +723,44 @@ FunctionEnd
...
+  !insertmacro StripSlash DATADIR
   DetailPrint "Checking whether an existing database needs upgrade: "
+  push ${ARANGO_EXIT_ALREADY_RUNNING}
+  pop $0
+retryDetectUpgradeNeeded:
   ExecWait "$INSTDIR\${SBIN_DIR}\arangod.exe --server.rest-server false --log.foreground-tty false --database.check-version" $0
   DetailPrint "done Checking whether an existing database needs upgrade: $0"
+  ${If} $0 == ${ARANGO_EXIT_ALREADY_RUNNING}
+  ${OrIf} $0 == ${ARANGO_EXIT_COULD_NOT_LOCK}
+    MessageBox MB_RETRYCANCEL "Another ArangoDB is currently locking your database, please stop it, and then click retry." IDCANCEL abortUpgrade IDRETRY retryDetectUpgradeNeeded
```
Comment (Simran-B) [inline Installation/Windows/Templates/NSIS.template.in]:
```suggestion
    MessageBox MB_RETRYCANCEL "Another ArangoDB server process is currently locking your database, please end it and then click retry." IDCANCEL abortUpgrade IDRETRY retryDetectUpgradeNeeded
```

File: Installation/Windows/Templates/NSIS.template.in [hunk 5]
```diff
@@ -715,36 +723,44 @@ FunctionEnd
...
+      DetailPrint "to backup directory $BackupPath"
+      !insertmacro xCopyDir "$DATADIR" "$BackupPath"
     ${EndIf}
     DetailPrint "Attempting to run database upgrade: "
     ; Now actually do the upgrade
     ExecWait "$INSTDIR\${SBIN_DIR}\arangod.exe --server.rest-server false  --log.level error --database.auto-upgrade true" $0
     DetailPrint "Done running database upgrade: $0"
     ${If} $0 != 0
-      !insertmacro printExitCode $0 "the Upgrade failed, please do a manual upgrade"
+      !insertmacro printExitCode $0 "the Upgrade failed, please do a manual upgrade" ""
```
Comment (Simran-B) [inline Installation/Windows/Templates/NSIS.template.in]:
```suggestion
      !insertmacro printExitCode $0 "The upgrade failed, please do a manual upgrade" ""
```

**Comment by dothebart** [comment]
Files on the installation root happenes now and then when nsis goes haywire during the xtraction process; We didn't find reproducible indicators when this would happen, how to trigger this, and how to fix it in NSIS. After that untracked files are on the installation root, and uninstall doesn't remove them. This happenes with all current installers and is not specific to this PR.

**Pull request merged**